sim-lib/src/lib.rs

```diff
@@ -825,72 +904,51 @@ async fn produce_events<N: DestinationGenerator + ?Sized, A: PaymentGenerator +
                     },
                 };
 
                 log::debug!("Generated random payment: {source} -> {}: {amount} msat.", destination);
 
                 // Send the payment, exiting if we can no longer send to the consumer.
                 let event = SimulationEvent::SendPayment(destination.clone(), amount);
-                if let Err(e) = sender.send(event).await {
-                    log::debug!(
-                        "Stopped random producer for {amount}: {source} -> {destination}. Consumer error: {e}.",
-                    );
-                    break;
+                if sender.send(event.clone()).await.is_err() {
+                    return Err(SimulationError::MpscChannelError (format!("Stopped random producer for {amount}: {source} -> {destination}.")));
                 }
+
             },
         }
     }
-
-    log::debug!("Stopped random activity producer {source}.");
-    shutdown.trigger();
 }
 
 async fn consume_simulation_results(
-    logger: Arc<Mutex<PaymentResultLogger>>,
-    receiver: Receiver<(Payment, PaymentResult)>,
-    listener: Listener,
-    write_results: Option<WriteResults>,
-) {
-    log::debug!("Simulation results consumer started.");
-
-    if let Err(e) = write_payment_results(logger, receiver, listener, write_results).await {
-        log::error!("Error while reporting payment results: {:?}.", e);
-    }
-
-    log::debug!("Simulation results consumer exiting.");
-}
-
-async fn write_payment_results(
     logger: Arc<Mutex<PaymentResultLogger>>,
     mut receiver: Receiver<(Payment, PaymentResult)>,
     listener: Listener,
     write_results: Option<WriteResults>,
 ) -> Result<(), SimulationError> {
-    let mut writer = write_results.and_then(|write_result| {
-        let file = write_result.results_dir.join(format!(
-            "simulation_{:?}.csv",
-            SystemTime::now()
-                .duration_since(SystemTime::UNIX_EPOCH)
-                .unwrap()
-                .as_secs()
-        ));
-        let writer = WriterBuilder::new().from_path(file).ok()?;
-        Some((writer, write_result.batch_size))
-    });
+    let mut writer = match write_results {
+        Some(res) => {
+            let duration = SystemTime::now().duration_since(SystemTime::UNIX_EPOCH)?;
+            let file = res
+                .results_dir
+                .join(format!("simulation_{:?}.csv", duration));
+            let writer = WriterBuilder::new().from_path(file)?;
+            Some((writer, res.batch_size))
+        },
+        None => None,
+    };
 
     let mut counter = 1;
+
     loop {
-        tokio::select! {
+        select! {
             biased;
             _ = listener.clone() => {
-                log::debug!("Simulation results consumer received shutdown signal.");
-                return writer.map_or(Ok(()), |(ref mut w, _)| w.flush().map_err(|_| SimulationError::FileError))
+                return Ok(());
```

Comment:
Still need to flush writer to disk otherwise pending items in the batch won't be written to disk?